list.py

```diff
@@ -78,33 +81,24 @@ def __getitem__(self, index):
 
 
     def __setitem__(self, index, elem):
-        self.check_index_borders(index)
-        self.find_node(self.index_transform(index)).elem = elem
+        index = self.check_index_and_transform(index)
+        self.find_node(index).elem = elem
         
 
     def __iter__(self):
         return ListIterator(self.head)
 
 
-    def index_transform(self, index: int) -> int:
-        '''
-        Transform negative indices to correct positive value
-        '''
-        return index + self.length if index < 0 else index
-
-
-    def check_index_borders(self, index, border = 1):
+    def check_index_and_transform(self, index, include_border = 0):
         '''
-        check index in: 
-            [-len; len): border = 1 (default)
-            [-len; len]: border = 0
+        [-len; len): border = 0, default.
+        [-len; len]: border = 1, for List.push and Queue.put;
         '''
-        if index > self.length - border or index < -self.length:
```

Comment:
`self.length **-** border`

Reply:
If you needed to note something, it's best to put that as a line comment `#` in code. The comments in Pull requests are best suited for ultimate code review

Reply:
Good. I misunderstood what the comments in pull request are for.

Reply:
Correct. If you want to leave a **persistent** (in other words: long-term) comment for the implementation - better use the code comments (the ones that begin with `#`).

If you wanted to discuss something or ask a question while making a PR it's also alright to use PR comments, but this particular comment at the top of the thread is too succinct. I don't understand what you wanted to say by that comment.

Anyhow, the commit message or PR description is also a place where it's possible to leave the short-term implementation comments that you expect for the people to see only once and never again.

There are now string rules as to what should be written where. However, I propose the following *convention*. This isn't a prescription of how every developer must write code, but these are the rules I use myself when commenting.

- Use code comments for persistent description of the code. This includes:
    - Documentation, which is a human-readable, general description of what the code does. Some languages (like python, don't use comments for that, but have own syntax like a block enclosed with `"""`):
      - [example 1 (library/module documentation)](https://github.com/Veetaha/snowpity/blob/299f45585f9d72117b62f0eba2c7634c50f9d1ce/metrics-bat/src/lib.rs#L1-L60)
      - [example 2 (function/method documentation)](https://github.com/Veetaha/snowpity/blob/299f45585f9d72117b62f0eba2c7634c50f9d1ce/snowpity-tg/src/media_host/twitter/rpc.rs#L151-L164)
	- If the code was copy-pasted, leave full and stable URL links with the references to the source:
	  - [example 1](https://github.com/Veetaha/snowpity/blob/299f45585f9d72117b62f0eba2c7634c50f9d1ce/metrics-bat/src/timing.rs#L20-L21)
	- If the code was based on an article, blog post, or even a StackOverflow answer, then leave the URL link to them
	  - [example 1](https://github.com/Veetaha/snowpity/blob/299f45585f9d72117b62f0eba2c7634c50f9d1ce/snowpity-tg/src/db/tg_chat/repo.rs#L96)
	  - [example 2](https://github.com/Veetaha/snowpity/blob/299f45585f9d72117b62f0eba2c7634c50f9d1ce/snowpity-tg/src/media_host/twitter/rpc.rs#L126)
	- Explanations of why the code is written this way and why we didn't choose to do something another way
	  - [example 1](https://github.com/Veetaha/snowpity/blob/54ce38375ca78e0fe52e6fbf598c87f572b21fc9/snowpity-tg/src/tg/captcha.rs#L224)
	  - [example 2](https://github.com/Veetaha/snowpity/blob/299f45585f9d72117b62f0eba2c7634c50f9d1ce/snowpity-tg/src/sysinfo.rs#L91-L96)
	- Explaining unobvious things:
	  - [example 1](https://github.com/Veetaha/snowpity/blob/299f45585f9d72117b62f0eba2c7634c50f9d1ce/metrics-bat/src/lib.rs#L266-L267)
	  - [example 2](https://github.com/Veetaha/snowpity/blob/299f45585f9d72117b62f0eba2c7634c50f9d1ce/metrics-bat/src/lib.rs#L249-L250)
	- `XXX:` comments for explaining super-unobvious things, where doing something wrong may lead to extreme confusion and potentially wasting hours of time trying to debug the program, or breaking something without noticing it:
	  - [example 1](https://github.com/Veetaha/snowpity/blob/54ce38375ca78e0fe52e6fbf598c87f572b21fc9/snowpity-tg/src/tg/inline_query.rs#L209-L210)
	  - [example 2](https://github.com/Veetaha/snowpity/blob/299f45585f9d72117b62f0eba2c7634c50f9d1ce/metrics-bat/src/lib.rs#L112-L113)
	- `FIXME:` comments to draw attention to things that may not work just yet, and we need to fix them:
	  - [example 1](https://github.com/Veetaha/snowpity/blob/54ce38375ca78e0fe52e6fbf598c87f572b21fc9/snowpity-tg/src/tg/media_cache/tg_upload.rs#L109)
	  - [example 2](https://github.com/Veetaha/snowpity/blob/54ce38375ca78e0fe52e6fbf598c87f572b21fc9/snowpity-tg/src/tg/media_cache/tg_upload.rs#L109)
	- `TODO:` comments to leave notes to yourself for the things you need to finish as part of the pull request, if you want to do something a bit later. ⚠️ Pay attention that `TODO:` must be removed before merging the PR. You are supposed to search for all `TODO:` comments in the project and implement what was deferred. Sometimes, such comments mistakenly get into the `master` branch (which is wrong), but because mistakes happen we do see such comments in `master` from time to time, like in the examples:
	  - [example 1](https://github.com/Veetaha/snowpity/blob/54ce38375ca78e0fe52e6fbf598c87f572b21fc9/snowpity-tg/src/tg/mod.rs#L172-L173) I forgot to do this before merging the PR
	  - [example 2](https://github.com/Veetaha/snowpity/blob/a67dd92fcff9fb8262ad1e70080d89a64fa9186f/snowpity-tg/src/tg/cmd/mod.rs#L30) I forgot to remove the `TODO` comment, because I actually did what is described in the comment. Stale comments, comments that lie, or are misleading are a big yet-unsolved problem of humanity. When we write code comments we need to make sure we keep them up to date and remove them when they are no longer useful *manually*.
- Use the commit message and the PR description (If you are using the "Squash and merge", I recommend copying the PR description into the squashed commit message) to describe the implemented changes in the commit or PR at a high level. This serves as an introduction to viewing the code diff (files changed) by the person reading the repository history. People generally don't read the repository history, the do that only if they want to understand why the code was written this way, and how we got to this point where it's written as this today. Then you would look through commits, and PRs, look at their messages and PR descriptions to understand the history of changes and what each commit and PR changes in general without looking at the code changes in them (there may be a lot of code changes, but you can describe them usually in a few words).
- Use PR comments to discuss the work-in-progress implementation
  - If you are the author of the PR, you may write your own comment threads if you want to consult with the reviewers, or ask them to pay attention to "this place" in code when reviewing if you want to hear feedback about some particular block of code
  - If you are the reviewer, you just put PR comments everywhere you think is reasonable: to also draw attention to potential problems (bugs) in code, give a hint on how something could be implemented better, share knowledge, etc.

Reply:
I copied it to myself. Very detailed and therefore understandable, thank you!